### PR TITLE
Helm chart now allows to enable TLS and basic auth

### PR DESCRIPTION
**Pull request by nvvfedorov**
This PR adds the ability to enable TLS and HTTP Basic authentication for dcgm-exporter. 

The following configuration values were added:

```yaml
# HTTPS configuration
tlsServerConfig:
  # Enable or disable HTTPS configuration
  enabled: false
  # Use autogenerated self-signed TLS certificates
  autoGenerated: true
  # Existing secret containing your own server key and certificate
  existingSecret: ""
  # Certificate file name
  certFilename: "tls.crt"
  # Key file name
  keyFilename: "tls.key"
  # CA certificate file name
  caFilename: "ca.crt"
  # Server policy for client authentication. Maps to ClientAuth Policies.
  # For more detail on clientAuth options:
  # https://golang.org/pkg/crypto/tls/#ClientAuthType
  #
  # NOTE: If you want to enable client authentication, you need to use
  # RequireAndVerifyClientCert. Other values are insecure.
  clientAuthType: ""
  # TLS Key for HTTPS - ignored if existingSecret is provided
  key: ""
  # TLS Certificate for HTTPS - ignored if existingSecret is provided
  cert: ""
  # CA Certificate for HTTPS - ignored if existingSecret is provided
  ca: ""

basicAuth:
  #Object containing <user>:<passwords> key-value pairs for each user that will have access via basic authentication
  users: {}
```

Also, new E2E tests were added:
- The dcgm-exporter deployment when TLS is enabled;
- The dcgm-exporter deployment when HTTP basic authentication is enabled.

The E2E suite now allows to specify which tests to run:
- `make e2es-tests`  - run all e2e tests;
- `make e2e-basic-auth` - run e2e scenario when basic auth is enabled;
- `make e2e-tls` - run e2e scenario when TLS is enabled;
- `make e2e-default` - run e2e scenario when tests deploy the helm chart with the default configuration.

### CODE REVIEW

File: tests/e2e/e2e_suite_test.go [hunk 1]
```diff
@@ -57,240 +53,61 @@ type testContextType struct {
...
 
-			_, _ = fmt.Fprintln(GinkgoWriter, "Clean up: completed")
-		})
+		helmClient = shouldCreateHelmClient(k8sConfig)
 
-		It("should create namespace", func(ctx context.Context) {
-			shouldCreateNamespace(ctx, kubeClient, labels)
+		BeforeAll(func(ctx context.Context) {
+			shouldCreateNamespace(ctx, kubeClient, testRunLabels)
 		})
```
Comment (rohit-arora-dev) [inline tests/e2e/e2e_suite_test.go:90]: Just out of curiosity why are these not part of `BeforeAll` anymore?

File: tests/e2e/main_test.go [hunk 2]
```diff
@@ -63,6 +63,11 @@ func TestMain(m *testing.M) {
 		"",
 		`DCGM-exporter command line arguments. Example: -arguments="{-f=/etc/dcgm-exporter/default-counters.csv}"`)
 
+	flag.BoolVar(&testContext.noCleanup,
+		"no-cleanup",
+		false,
+		`Skip clean up after tests execution`)
+
```
Comment (rohit-arora-dev) [inline tests/e2e/main_test.go:70]: In what scenario we need this flag?

**Review by rohit-arora-dev — COMMENTED** [review]
Just a couple of minor things, looking god otherwise.

File: deployment/templates/daemonset.yaml [hunk 3]
```diff
@@ -78,6 +78,18 @@ spec:
...
+        secret:
+          secretName: {{ include "dcgm-exporter.tlsCertsSecretName" . }}
+          defaultMode: 0664
+    {{- end }}
+    {{- if or .Values.tlsServerConfig.enabled $.Values.basicAuth.users}}
+      - name: "web-config-yaml"
+        configMap:
+          name: {{ include "dcgm-exporter.webConfigConfigMap" . }}
+          defaultMode: 0664
+    {{- end }}
```
[the remark this answers is not in the document]
Comment (nvvfedorov) [reply deployment/templates/daemonset.yaml:92]: @rohit-arora-dev, can you be more specific about where the issue is?

[the remark this answers is not in the document]
Comment (rohit-arora-dev) [reply deployment/templates/daemonset.yaml:92]: Please look at the file, it has indentation issues.

[the remark this answers is not in the document]
Comment (rohit-arora-dev) [reply deployment/templates/daemonset.yaml:92]: Line 81-92

File: tests/e2e/e2e_suite_test.go (hunk 1 above)
Comment (nvvfedorov) [reply tests/e2e/e2e_suite_test.go:90]: @rohit-arora-dev  Is it an old comment?  BeforeAll now creates a namespace as part of the BeforeAll. 

Global variables such as kubeClient, helmClient, and testRunLabels initialize once in the Context, which is the container (https://onsi.github.io/ginkgo/#organizing-specs-with-container-nodes. ) for e2e tests. After global variables are initialized, they are available for child specs, like BeforeAll, etc.

File: tests/e2e/main_test.go (hunk 2 above)
Comment (nvvfedorov) [reply tests/e2e/main_test.go:70]: @rohit-arora-dev Troubleshooting. E2E destroys all resources after the end of the test or cancelation. The flag allows to keep created resources for the investigation of failures.

File: deployment/templates/daemonset.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (nvvfedorov) [reply deployment/templates/daemonset.yaml:92]: @rohit-arora-dev  Why do you think indentation is wrong when K8S accepts this yaml without issues?
SCR-20240403-onbs

[the remark this answers is not in the document]
Comment (nvvfedorov) [reply deployment/templates/daemonset.yaml:92]: If your question is about the `if` - `end` blocks, it is made intentionally to make it accessible to see when the block starts and ends.

[the remark this answers is not in the document]
Comment (rohit-arora-dev) [reply deployment/templates/daemonset.yaml:92]: I see a mix of styles, the file in general follow a style where if-end aligns with the indentation, and does not stand out. Even in other parts of the code where you have added new if-end blocks the indentation is correct, but for this specific block it is different.

My suggestion is to make it consistent across the file, either change all the if-end to follow this style or this block should follow the style consistent with the rest of the file.

[the remark this answers is not in the document]
Comment (nvvfedorov) [reply deployment/templates/daemonset.yaml]: @rohit-arora-dev  done.

[the remark this answers is not in the document]
Comment (rohit-arora-dev) [reply deployment/templates/daemonset.yaml]: Thanks, definitely looking better.

**Review by rohit-arora-dev — APPROVED** [review]
/LGTM

**Pull request closed without merging**